docs/modules/ROOT/pages/upgrades.adoc

```diff
@@ -68,46 +65,29 @@ mod UpgradeableContract {
-        Upgradeable::InternalImpl::upgrade(ref upgradeable_state, new_class_hash);
+    impl UpgradeableImpl of IUpgradeable<ContractState> {
+        fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
+            // This function can only be called by the owner
+            let ownable_state = Ownable::unsafe_new_contract_state();
+            Ownable::InternalImpl::assert_only_owner(@ownable_state);
+
+            // Replace the class hash upgrading the contract
+            let mut upgradeable_state = Upgradeable::unsafe_new_contract_state();
+            Upgradeable::InternalImpl::_upgrade(ref upgradeable_state, new_class_hash);
+        }
     }
 
     (...)
 }
 ----
 
-=== Interface
-
-Implementing this interface to support upgradeability, supports interoperability by for example allowing DAOs
-and plugins to connect to contracts and upgrade them.
-
-[,javascript]
-----
-trait IUpgradeable {
-    fn upgrade(new_class_hash: ClassHash);
-}
-----
-
 == Proxies and Starknet
 
-A proxy contract is a contract that delegates function calls to another contract.
-This type of pattern decouples state and logic.
-Proxy contracts store the state and redirect function calls to an implementation contract that handles the logic,
-usually catching these calls with a fallback mechanism.
-This allows for different patterns such as upgrades, where implementation contracts can change but the proxy contract (and thus the state) does not;
-as well as deploying multiple proxy instances pointing to the same implementation.
-This can be useful to deploy many contracts with identical logic but unique initialization data.
+Proxies allow for different patterns such as upgrades as well as deploying multiple clones of the same contract. But Starknet achieves these two things in different ways.
 
-In the case of contract upgrades, it is achieved by simply changing the proxy's reference to the class hash of the declared implementation.
-This allows developers to add features, update logic, and fix bugs without touching the state or the contract address to interact with the application.
+In the case of contract upgrades, it is achieved by simply changing the contract's class hash. As of clones, contracts already are like clones of the class they implement.
 
-In Starknet, implementing a proxy pattern has an important limitation: there is no fallback mechanism to be used
-for redirecting every potential function call to the implementation contract. This means a generic proxy contract
+Implementing a proxy pattern in Starknet has an important limitation: there is no fallback mechanism to be used
+for redirecting every potential function call to the implementation. This means that a generic proxy contract
 can't be implemented. Instead, a proxy contract must be implemented with each specific function that is going to be redirected.
 This can still be useful for example with upgrading the logic of some functions.
```

Comment:
```suggestion
can't be implemented. Instead, a limited proxy contract can implement specific functions that forward their execution to another contract class.
This can still be useful for example to upgrade the logic of some functions.
```
i think this reads better, but still not convinced about documenting a half baked proxy design

Reply:
IMO we don't have a half baked proxy design, there are multiple potential proxy designs, and we don't have one, but we have the ability to implement them, proxies can be implemented in Starknet (different patterns).

Reply:
> I think having a fallback makes proxies easier to implement, but as with the examples I shared, we can implement proxies without it

agree

> I wouldn't call them limited proxies

I would since there's strong limitations as not being able to support arbitrary function forwarding (fallback function).

> IMO we don't have a half baked proxy design, there are multiple potential proxy designs, and we don't have one, but we have the ability to implement them, proxies can be implemented in Starknet (different patterns).

exactly because we don't have one is that I called the proxy described as "half baked". It's a (very valid) assumption of how a proxy would look like but it's briefly explained, without a clear use case, and requires to be more thoroughly thought before being described/suggested in our official docs, as they're seen as a place to learn about best practices.